feat: sprint11 add-database

dmitry-razumov opened the pull request:
feat: sprint 11 added database, mpa, genre & etc

dmitry-razumov committed:
feat: sprint11 add-database

dmitry-razumov committed:
fix:checkstyle  del unused import

dmitry-razumov committed:
fix:checkstyle

dmitry-razumov committed:
fix:filmControllerTest, added Mpa

dmitry-razumov committed:
fix:checkstyle

dmitry-razumov committed:
fix:schema.sql

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/controller/GenreController.java:
```diff
@@ -0,0 +1,35 @@
+@RequiredArgsConstructor
+@Slf4j
+@Validated
+public class GenreController {
```
Правильное решение выделить отдельные классы-контроллеры для жанра и рейтинга.

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -28,6 +26,13 @@ public class Film {
     private long duration;
-    Set<Long> likes = new HashSet<>();
     private int rate = 0;
+    private Set<Long> likes = new HashSet<>();
```
Коллекция лайков более не требуется, так как хранение мы осуществляем в БД)
А для пользователя коллекция с друзьями

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/model/Genre.java:
```diff
@@ -0,0 +1,15 @@
+public class Genre {
+    private final int id;
+    @NotBlank
+    private final String name;
```
Почему финализировал поля?

dmitry-razumov replied:
забыл убрать, остатки былых измышлений

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/model/Mpa.java:
```diff
@@ -0,0 +1,15 @@
+public class Mpa {
+    private final int id;
+    @NotBlank
+    private final String name;
```
И здесь тоже

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -4,8 +4,10 @@
+import java.util.stream.Collectors;
 
 @Component
 @Slf4j
```
Ниже напротив класса написать не могу) Эта реализация более не будет использовать, ее можно исключить) Также отпадает потребность в @qualifier

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
```diff
@@ -2,13 +2,15 @@
 @Component
 @Slf4j
+@Qualifier("inMemoryUserStorage")
 public class InMemoryUserStorage implements UserStorage {
```
Ниже напротив класса написать не могу) Эта реализация более не будет использовать, ее можно исключить) Также отпадает потребность в @qualifier

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/validator/exist/ExistValidator.java:
```diff
@@ -1,21 +1,38 @@
 
 @Slf4j
-@RequiredArgsConstructor
+//@RequiredArgsConstructor
```
Закомментированный код следует удалять

dmitry-razumov replied:
ок, но вернул обратно поскольку @qualifier  больше не требуется

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/validator/exist/ExistValidator.java:
```diff
@@ -1,21 +1,38 @@
+    private final FilmStorage filmStorage;
+    private final MpaStorage mpaStorage;
+    private final GenreStorage genreStorage;
     String type;
```
Где модификатор потерял?)

TarasovSam commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,50 @@
+DROP TABLE IF EXISTS film_genre;
+DROP TABLE IF EXISTS film_likes;
+
+CREATE TABLE IF NOT EXISTS users (
```
Можно было бы ввести ограничение на уникальность логина и почты на уровне таблицы в БД)

create unique index if not exists USER_EMAIL_UINDEX on USERS (email);
create unique index if not exists USER_LOGIN_UINDEX on USERS (login);

TarasovSam reviewed (commented):
Дмитрий, привет!
Проверил твое задание, все довольно хорошо. Оставил комментарии с тем, что можно улучшить и что нужно поправить, плюс пару вопросов)
Жду твою работу на следующую итерацию!) Если будут вопросы, пиши в пачку)

dmitry-razumov committed:
refact: удалил сеты likes, friends, поправил методы get userFriends, commonFriends
fix: remove final of fields Genre, Mpa
refact: удалил inMemory storage's, @qualifier's и конструкторы по коду
fix: add модификатор в existValidator
add: добавил констрейны на уникальность в схему бд на user email и login

dmitry-razumov committed:
fix:checkstyle

dmitry-razumov committed:
fix:checkstyle

dmitry-razumov committed:
fix:checkstyle

TarasovSam commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -3,22 +3,19 @@
-    public FilmService(@Qualifier("filmDbStorage") FilmStorage filmStorage) {
-        this.filmStorage = filmStorage;
-    }
-
```
Блин, как я пропустил в прошлый раз...
Строка 61 - предварительно следует проверить на null, иначе можно словить NPE

dmitry-razumov replied:
Верно ли я понимаю, что речь за NPE при film.getReleaseDate()? а разве недостаточно @Valid в контроллере и @NotNull в сущности надо полем?
public class Film {
....
    @NotNull(message = "Дата релиза фильма не должна быть пустой")
    private LocalDate releaseDate;
...

TarasovSam reviewed (approved):
Юлия, проверил твое задание, спасибо за работу и за то, что дорабатываешь код. Оставил маленький комментарий))
Задание я принимаю, успеха в обучении и до следующего спринта!) Если будут вопросы, пиши в пачку)
